File: solutions/javascript/proverb/1/proverb.js

export function proverb(...words) {

  const qualifier = (typeof words[words.length-1] == 'object') ? `${words.pop().qualifier} ` : '';
  if (words.length == 0) return '';
  
  let proverb = '';
  let previousWord = words[0];
  for (const word of words.slice(1)) {
    proverb += `For want of a ${previousWord} the ${word} was lost.\n`;
    previousWord = word;
  }
  proverb += `And all for the want of a ${qualifier}${words[0]}.`;
  return proverb;
};

File: solutions/elixir/proverb/1/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""
  def recite(strings=[first|others]) do
    sentences = Enum.zip(strings, others)
        |> Enum.map(fn {f,s} -> "For want of a #{f} the #{s} was lost." end)
    Enum.join(sentences ++ ["And all for the want of a #{first}.\n"], "\n")
  end
end
